File: 卡码/代码随想录/C++/链表/03-反转链表/LeetCode_0206.cpp
//206. 反转链表
#include <iostream>
using namespace std;

struct ListNode{
	int val;
	ListNode* next;
	ListNode():val(0),next(nullptr){}
	ListNode(int val):val(val),next(nullptr){}
	ListNode(int val,ListNode* next):val(val),next(next){}
};

class Solution {
public:
	//递归
	ListNode* reverseList(ListNode* head){
		//如果链表为空或者只有一个元素
		if(head==nullptr||head->next==nullptr) return head;
		ListNode* node=reverseList(head->next);
		head->next->next=head;
		head->next=nullptr;
		return node;
	}
	//双指针法
  ListNode* reverseList1(ListNode* head) {
  	//如果链表为空或者只有一个元素
		if(head==nullptr||head->next==nullptr) return head;
		ListNode* pre=nullptr;
		ListNode* cur=head;
		while(cur){
			ListNode* tmp=cur->next;
			cur->next=pre;
			pre=cur;
			cur=tmp;
		}
		return pre;
  }
};

int main(){
	return 0;
}

File: 卡码/代码随想录/JS/链表/03-反转链表/LeetCode_0206.js
//206. 反转链表
function ListNode(val,next){
  this.val=(val===undefined ? 0 : val);
  this.next=(next===undefined ? null : next);
}

//双指针法
const reverseList1=function(head){
  //如果链表为空或者只有一个元素
  if(head==null||head.next==null) return head;
  let pre=null;
  let cur=head;
  while(cur){
    let tmp=cur.next;
    cur.next=pre;
    pre=cur;
    cur=tmp;
  }
  return pre;
}

//递归
const reverseList=head=>{
  //如果链表为空或者只有一个元素
  if(head==null||head.next==null) return head;
  let node=reverseList(head.next);
  head.next.next=head;
  head.next=null;
  return node;
}